src/crystal/system/win32/event_loop_iocp.cr

```diff
@@ -1,14 +1,30 @@
+require "c/iocp"
 require "crystal/system/print_error"
 
 module Crystal::EventLoop
-  @@queue = Deque(Fiber).new
+  @@queue = Deque(Crystal::Event).new
 
   # Runs the event loop.
   def self.run_once : Nil
-    next_fiber = @@queue.pop?
+    unless @@queue.empty?
+      next_event = @@queue.min_by { |e| e.wake_in }
+      time_elapsed = (Time.monotonic - next_event.slept_at)
 
-    if next_fiber
-      Crystal::Scheduler.enqueue next_fiber
+      unless time_elapsed > next_event.wake_in
+        sleepy_time = (next_event.wake_in - time_elapsed).total_milliseconds.to_i
+        io_entry = Slice.new(1, LibC::OVERLAPPED_ENTRY.new)
+
+        if LibC.GetQueuedCompletionStatusEx(Thread.current.iocp, io_entry, 1, out removed, sleepy_time, false)
+          if removed == 1 && io_entry.first.lpOverlapped
+            next_event = io_entry.first.lpOverlapped.value.cEvent.unsafe_as(Crystal::Event)
+          end
+        else
+          raise RuntimeError.from_winerror("Error getting i/o completion status")
```

Comment:
@incognitorobito  I think you need to check the error code here since a timeout will trow an error in this case. See [GetQueuedCompletionStatusEx docs on the timeout parameter](https://docs.microsoft.com/en-us/windows/win32/fileio/getqueuedcompletionstatusex-func#parameters)

>The number of milliseconds that the caller is willing to wait for a completion packet to appear at the completion port. If a completion packet does not appear within the specified time, the function times out and returns FALSE.

I think you need to call `GetLastError` and check if the error code is `WAIT_TIMEOUT` (code 258). If that's the case you know it was either an event that timed out or a timer that has expired.